private/buf/cmd/buf/buf.go

```diff
@@ -317,14 +317,37 @@ func wrapError(err error) error {
 		connectCode := connectErr.Code()
 		switch {
 		case connectCode == connect.CodeUnauthenticated, isEmptyUnknownError(err):
-			if authErr, ok := bufconnect.AsAuthError(err); ok && authErr.TokenEnvKey() != "" {
+			authErr, ok := bufconnect.AsAuthError(err)
+			if !ok {
+				// This code should be unreachable.
+				return fmt.Errorf("Failure: you are not authenticated. "+
+					"Set the %[1]s environment varialbe or create a new entry in your netrc, "+
+					"using a Buf API Key as the password. If you have set the %[1]s or already have an entry in your netrc, check "+
+					"to see that your token is not expired. For details, visit https://buf.build/docs/bsr/authentication",
+					bufconnect.TokenEnvKey,
+				)
+			}
+			// Invalid token found in env var.
+			if authErr.HasToken() && authErr.TokenEnvKey() != "" {
 				return fmt.Errorf("Failure: the %[1]s environment variable is set, but is not valid. "+
-					"Set %[1]s to a valid Buf API key, or unset it. For details, "+
-					"visit https://docs.buf.build/bsr/authentication", authErr.TokenEnvKey())
+					"Set %[1]s to a valid Buf API key, or unset it. For details, visit https://buf.build/docs/bsr/authentication",
+					authErr.TokenEnvKey(),
+				)
+			}
+			// Invalid token found in netrc.
+			if authErr.HasToken() {
+				return fmt.Errorf("Failure: invalid token for %s found in netrc. "+
+					"Set it to a valid Buf API key. For details, visit https://buf.build/docs/bsr/authentication",
+					authErr.Remote(),
+				)
 			}
-			return errors.New("Failure: you are not authenticated. Create a new entry in your netrc, " +
-				"using a Buf API Key as the password. If you already have an entry in your netrc, check " +
-				"to see that your token is not expired. For details, visit https://docs.buf.build/bsr/authentication")
+			// No token found.
+			return fmt.Errorf("Failure: you are not authenticated for %s. "+
+				"Set the %s environment variable or create a new entry in your netrc, "+
```

Comment:
"Create a new entry in your netrc" isn't really how we want people to fix this problem, we want people to use `buf registry login [HOST]`, where you should be able to auto-populate `[HOST]` based on the hostname you used to call the API. if `[HOST]` is `buf.build`, omit it.

Reply:
I think [6f5be28](https://github.com/bufbuild/buf/pull/3093/commits/6f5be285eabb8e4e9820937d1a0bc9ad03a1967c) resolves it.

Reply:
yep1